### PR TITLE
build from source

### PR DESCRIPTION
**Pull request by mfschumann**

### CODE REVIEW

**Comment by mfschumann** [comment]
bot,build

**Comment by mfschumann** [comment]
bot, build

**Comment by mfschumann** [comment]
bot, build

**Comment by mfschumann** [comment]
bot, build

**Comment by mfschumann** [comment]
bot, build

**Comment by mfschumann** [comment]
bot, build

**Comment by Betterbird** [comment]
Added a tag for 102.2.2-bb16 now: https://github.com/Betterbird/thunderbird-patches/releases/tag/102.2.2-bb16

**Comment by mfschumann** [comment]
I'd say this pull request can be merged now:
- Betterbird is built from source for x86_64 and aarch64
- Language support works

Any further remarks or wishes @Betterbird or @hfiguiere ?

**Comment by Betterbird** [comment]
> Any further remarks or wishes @Betterbird or @hfiguiere ?

Remarks, yes, or more a question: So in general terms, you download the TB flat source, apply the BB patches, compile the lot, and then hack the language packs into place patching the TB XPIs. That's all understood. What interests me (and likely @dbonner) is how you set up the compile/build environment. I'm sure you've read our "manual" build instructions (https://github.com/Betterbird/thunderbird-patches#building-betterbird) which involve installing a few packages (git, various python, mercurial, rust) and then [our build script](https://github.com/Betterbird/thunderbird-patches/blob/main/build/build.sh) runs the Mozilla bootstrap which for X86-64 works and installs a whole heap of other packages (clang, node, cbindgen, etc.) to complete the build environment. Also, the build script sets the rust version since with a non-compatible rust compiler the build fails. Mozilla doesn't support aarch64, so their bootstrap doesn't work there and the [build instructions for aarch64](https://github.com/Betterbird/thunderbird-patches/blob/main/build/build-env-aarch64.MD) are more complicated. So the question is: How do you do it? Maybe you have all the expertise from building TB. Somewhere in the conversation (I believe) @hfiguiere said something like "we have aarch64" builders, but having the physical/virtual machine is just the start.

**Comment by mfschumann** [comment]
bot, build

File: eu.betterbird.Betterbird.json [hunk 1]
```diff
@@ -1,70 +1,298 @@
...
-                    "path": "eu.betterbird.Betterbird.appdata.xml"
-                }
-            ]
+  "id": "eu.betterbird.Betterbird",
+  "runtime": "org.freedesktop.Platform",
+  "runtime-version": "21.08",
+  "sdk": "org.freedesktop.Sdk",
+  "sdk-extensions": [
+    "org.freedesktop.Sdk.Extension.rust-stable",
+    "org.freedesktop.Sdk.Extension.node14",
```
Comment (hfiguiere) [inline eu.betterbird.Betterbird.json:8]: any chance to move to node16? It seems that node14 doesn't build with the 22.08 runtime.

not blocking at the moment.

File: flathub.json [hunk 2]
```diff
@@ -1,3 +1,3 @@
 {
```
Comment (hfiguiere) [inline flathub.json:1]: you can remove the whole file. no other arch is supported.

Comment (mfschumann) [reply flathub.json]: Done.

File: eu.betterbird.Betterbird.json (hunk 1 above)
Comment (mfschumann) [reply eu.betterbird.Betterbird.json:8]: Frankly I have no idea where Thunderbird/Betterbird relies on node.js and if that part will be compatible with node16. Maybe @Betterbird can comment on that?

I don't really see any downside of staying with freedesktop 21.08 for now, so the pressure for upgrading to node16 is rather low I'd say.

**Comment by Betterbird** [comment]
Node.js is used in the Mozilla platform code, not TB/BB code. I can't tell what they're doing with it exactly, one post on a Firefox mailing list in 2019 said:
_Now that we have NodeJS in the tree, we want to unlock the rich set of tooling in the NodeJS/npm ecosystem so that we can start depending on things available there in a coherent, managed, more secure way._

**Comment by mfschumann** [comment]
> ... So the question is: How do you do it? Maybe you have all the expertise from building TB. Somewhere in the conversation (I believe) @hfiguiere said something like "_we have aarch64 builders_", but having the physical/virtual machine is just the start.

Setting up the build environment is one of the things done by `flatpak-builder`, the tool that reads the "manifest" `eu.betterbird.Betterbird.json`. The largest part of the manifest used for Betterbird is taken from the Thunderbird manifest. So I did not have to fiddle with setting up the correct build environment and getting the dependencies right myself.

Many of the build dependencies for Betterbird are already available in the freedesktop runtime, the freedesktop SDK, or an SDK extension (e.g. rust). The rest (just a handful) is built from source before building Betterbird. All of this is controlled by the manifest and is agnostic of the platform, so you should be able to build the flatpak on any machine by cloning this git repo, installing flatpak-builder, and running `flatpak-builder build-dir eu.betterbird.Betterbird.json`. This feature of flatpak made building on aarch64 just work after setting everything up for x86_64.

In case you want to learn more about how flatpaks are built, [this documentation](https://docs.flatpak.org/en/latest/building.html) is a good starting point.

File: eu.betterbird.Betterbird.json (hunk 1 above)
Comment (mfschumann) [reply eu.betterbird.Betterbird.json:8]: Taking into account Betterbird's comment on node.js being part of the Mozilla platform I think it is out of scope of this flatpak to test if Thunderbird/Betterbird works with node16. So I will keep the runtime 21.08 and node14 until the Thunderbird flatpak is upgraded.

**Pull request merged**